### ISSUE TITLE
Multiple laser scanners with per-point time are not unwarped

### ISSUE THREAD

**Issue by gaschler** [googlecartographer/cartographer#910]
For multiple laser scans that have per-point timestamps, the following check fails:
```
[ INFO] [1518779619.525013825, 1405076296.631876676]: I0216 12:13:39.000000 31015 local_trajectory_builder.cc:103] Extrapolator is still initializing.
F0216 12:13:39.525270 31015 pose_extrapolator.cc:229] Check failed: time >= imu_tracker->time() (635406730966328436 vs. 635406730966334607) 
[FATAL] [1518779619.526051032, 1405076296.631876676]: F0216 12:13:39.000000 31015 pose_extrapolator.cc:229] Check failed: time >= imu_tracker->time() (635406730966328436 vs. 635406730966334607) 
    @           0x579513  cartographer::mapping::PoseExtrapolator::ExtrapolateRotation()
    @           0x579eab  cartographer::mapping::PoseExtrapolator::ExtrapolatePose()
    @           0x5b4427  cartographer::mapping_2d::LocalTrajectoryBuilder::AddRangeData()
    @           0x580cf0  cartographer::mapping::GlobalTrajectoryBuilder<>::AddSensorData()
    @           0x58305b  cartographer::mapping::CollatedTrajectoryBuilder::HandleCollatedSensorData()
    @           0x583181  _ZNSt17_Function_handlerIFvRKSsSt10unique_ptrIN12cartographer6sensor4DataESt14default_deleteIS5_EEEZNS3_7mapping25CollatedTrajectoryBuilderC1EPNS4_17CollatorInterfaceEiRKSt3setINSA_26TrajectoryBuilderInterface8SensorIdESt4lessISG_ESaISG_EES2_ISF_S6_ISF_EEEUlS1_S8_E_E9_M_invokeERKSt9_Any_dataS1_S8_
    @           0x620d9b  _ZNSt17_Function_handlerIFvSt10unique_ptrIN12cartographer6sensor4DataESt14default_deleteIS3_EEEZNS2_8Collator13AddTrajectoryEiRKSt13unordered_setISsSt4hashISsESt8equal_toISsESaISsEERKSt8functionIFvRKSsS6_EEEUlS6_E_E9_M_invokeERKSt9_Any_dataS6_
```

How to reproduce:
Download https://storage.googleapis.com/cartographer-public-data/bags/backpack_2d/b0-2014-07-11-10-58-16.bag.
Modify backpack_2d.lua as follows:
```
  num_multi_echo_laser_scans = 2,
```
Run launch file:
```
<launch>
  <param name="/use_sim_time" value="true" />
  <param name="robot_description"
    textfile="$(find cartographer_ros)/urdf/backpack_2d.urdf" />
  <node name="robot_state_publisher" pkg="robot_state_publisher"
    type="robot_state_publisher" />
  <node name="cartographer_node" pkg="cartographer_ros"
      type="cartographer_node" args="
          -configuration_directory $(find cartographer_ros)/configuration_files
          -configuration_basename backpack_2d.lua"
      output="screen">
    <remap from="echoes_1" to="horizontal_laser_2d" />
    <remap from="echoes_2" to="vertical_laser_2d" />
  </node>
  <node name="cartographer_occupancy_grid_node" pkg="cartographer_ros"
      type="cartographer_occupancy_grid_node" args="-resolution 0.05" />
  <node name="rviz" pkg="rviz" type="rviz" required="true"
      args="-d $(find cartographer_ros)/configuration_files/demo_2d.rviz" />
  <node name="playbag" pkg="rosbag" type="play"
      args="--clock $(env HOME)/Downloads/b0-2014-07-11-10-58-16.bag" />
</launch>
```

**Comment by gaschler** [comment]
Here is a work around (not a real fix): https://github.com/googlecartographer/cartographer/pull/911

**Comment by MichaelGrupp** [comment]
That's funny - I had to deal with the **exact same issue** yesterday - here' s my version of a quick fix: https://github.com/magazino/cartographer/commit/9f14301227ce525d32d49c6bfe26e447481127fc

**Comment by gaschler** [comment]
Update: After the related fix https://github.com/googlecartographer/cartographer/pull/936, the check no longer fails.
But, time-overlapping points will only be unwarped with the latest extrapolator time (not the exact per-point time), which is not optimal, and you see a warning for each overlapping PointCloud.

**Comment by gaschler** [comment]
Fixed by #996, closing.
